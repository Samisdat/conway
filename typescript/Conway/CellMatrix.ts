import {Cell} from './cell';
import {Position} from './position';

interface CellMatrixInterface {
    [index: string]: Cell;
}

export class CellMatrix {

    private matrix: CellMatrixInterface = {};

    public get(position: Position): Cell | undefined {

        if (undefined !== this.matrix[position.toString()]) {
            return this.matrix[position.toString()];
        }

        return undefined;


    }

    public add(cell: Cell): void {

        if (undefined !== this.matrix[cell.position.toString()]) {
            console.log('argh');
        }

        this.matrix[cell.position.toString()] = cell;

    }

    public remove(cell: Cell): void {

        if (undefined !== this.matrix[cell.position.toString()]) {
            delete this.matrix[cell.position.toString()];
        }

    }

    public all(): Cell[] {

        const cells: Cell[] = [];

        for (const index in this.matrix) {

            cells.push(this.matrix[index]);

        }

        return cells;

    }


}
