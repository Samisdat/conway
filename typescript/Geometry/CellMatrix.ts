import {Position} from '@Conway/Geometry/Position';

interface CellMatrixInterface {
    [index: string]: Position;
}

interface Dimension {
    width: number;
    height: number;
}

export class CellMatrix {

    private matrix: CellMatrixInterface = {};

    public has(position: Position): boolean {

        return (undefined !== this.matrix[position.toString()]);

    }

    public add(position: Position): void {

        if (true === this.has(position)) {
            return;
        }

        this.matrix[position.toString()] = position;

    }

    public remove(position: Position): void {

        if (false === this.has(position)) {
            return;
        }

        delete this.matrix[position.toString()];

    }

    public all(): Position[] {

        const positions: Position[] = [];

        for (const index in this.matrix) {

            positions.push(this.matrix[index]);

        }

        return positions;

    }

    private getDimension(): Dimension {

        const dimension: Dimension = {
            width: 0,
            height: 0
        }

        let left = 0;
        let right = 0;

        let top = 0;
        let bottom = 0;

        for (const position of this.all()) {

            if (position.x < left) {
                left = position.x;
            }

            if (position.x > right) {
                right = position.x;
            }

            if (position.y < bottom) {
                bottom = position.y;
            }

            if (position.y > top) {
                top = position.y;
            }

        }

        if (0 > left && 0 < right) {
            dimension.width = right - left;
        }
        else{
            dimension.width = right - left;
        }

        if (0 > bottom && 0 < top) {
            dimension.height = top - bottom;
        }
        else{
            dimension.height = top - bottom;
        }

        return dimension;

    }

    public width(): number {

        return this.getDimension().width;

    }

    public height(): number {

        return this.getDimension().height;

    }

}
