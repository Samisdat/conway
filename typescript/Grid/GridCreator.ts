import { Position } from '../position';

export class GridCreator {


    public readonly canvasWidth:number;
    public readonly canvasHeigth:number;
    public readonly cellWidthAndHeight:number;
    public readonly pan:Position;

    private rows:number;
    private cols:number;

    constructor(
        canvasWidth:number,
        canvasHeigth:number,
        cellWidthAndHeight:number,
        pan:Position
    ) {
        this.canvasWidth = canvasWidth;
        this.canvasHeigth = canvasHeigth;
        this.cellWidthAndHeight = cellWidthAndHeight;
        this.pan = pan;


        this.rows = this.canvasWidth / cellWidthAndHeight;
        this.cols = this.canvasHeigth / cellWidthAndHeight;

    }

    public getRows():number{
        return this.rows;
    }

    public getCols():number{
        return this.cols;
    }

    public getPan():Position{
        return this.pan;
    }

}