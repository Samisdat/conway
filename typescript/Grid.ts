import { Position } from './Conway/position';
import { GridCell } from './Grid/GridCell';
import { Habitat } from './Conway/habitat';
import { CellTypeLiving } from 'CellType/CellTypeLiving';
import { CellTypesCenter } from 'CellType/CellTypesCenter';
import { GridDimension } from 'Grid/GridDimension';
import { IntegerPosition } from './Conway/IntegerPosition';
import {CellTypesFactory} from 'CellType/CellTypesFactory';

export class Grid {

    private readonly habitat: Habitat;

    private gridDimension: GridDimension;
    private sourcePosition: IntegerPosition;

    private offset: Position;

    private cells: GridCell[] = [];

    public readonly center: IntegerPosition;

    private absoluteMap: {} = {};

    constructor(
        habitat: Habitat,
        gridDimension: GridDimension,
        sourcePosition: IntegerPosition,
        offset: Position
    ) {

        this.ensureOffsetBetweenMinusOneAndOne(offset);

        this.habitat = habitat;

        this.gridDimension = gridDimension;

        this.sourcePosition = sourcePosition;

        this.offset = offset;

        // create rows

        let relativePointer = new IntegerPosition(0, 0);
        let absolutePointer = this.sourcePosition;

        this.center = relativePointer.clone().move(
            new IntegerPosition(
                Math.floor(this.gridDimension.rows / 2),
                Math.floor(this.gridDimension.cols / 2)
            )
        );

        relativePointer = relativePointer.move(
            new Position(
                -1 * Math.floor(this.gridDimension.rows / 2),
                -1 * Math.floor(this.gridDimension.cols / 2)
            )
        );

        absolutePointer = absolutePointer.move(
            new Position(
                -1 * Math.floor(this.gridDimension.rows / 2),
                -1 * Math.floor(this.gridDimension.cols / 2)
            )
        );

        const moveRight = new Position(
            1,
            0
        );

        // @TODO Use Bounds

        for (let top = 0; top < this.gridDimension.cols; top += 1) {

            for (let right = 0; right < this.gridDimension.rows; right += 1) {

                const gridCell = new GridCell(
                    relativePointer,
                    absolutePointer,
                    this.offset
                );

                const absolutePosition = gridCell.absolutePosition;

                if (undefined === this.absoluteMap[absolutePosition.x]) {
                    this.absoluteMap[absolutePosition.x] = {};
                }

                this.absoluteMap[absolutePosition.x][absolutePosition.y] = this.cells.length;

                this.cells.push(gridCell);

                relativePointer = relativePointer.move(
                    moveRight
                );

                absolutePointer = absolutePointer.move(moveRight);

            }

            relativePointer = relativePointer.move(
                new IntegerPosition(-1 * this.gridDimension.rows, 1),
            );

            absolutePointer = absolutePointer.move(
                new IntegerPosition(-1 * this.gridDimension.rows, 1),
            );

        }

        for (let positionWithLivingCells of habitat.get()) {

            if (undefined === this.absoluteMap[positionWithLivingCells.x]) {
                continue;
            }
            if (undefined === this.absoluteMap[positionWithLivingCells.x][positionWithLivingCells.y]) {
                continue;
            }

            const cellIndex = this.absoluteMap[positionWithLivingCells.x][positionWithLivingCells.y];

            this.cells[cellIndex].setType(CellTypesFactory.get().living());

        }

    }

    private ensureOffsetBetweenMinusOneAndOne(offset: Position) {

        if (1 < Math.abs(offset.x )) {
            throw new Error('offset.x must be between -1 and 1');
        }

        if (1 < Math.abs(offset.y)) {
            throw new Error('offset.y must be between -1 and 1');
        }


    }

    public getSourcePosition(): IntegerPosition {
        return this.sourcePosition;
    }

    public getOffset(): Position {
        return this.offset;
    }

    public getCols(): number {
        return this.gridDimension.cols;
    }

    public getRows(): number {
        return this.gridDimension.rows;
    }

    public getCells() {
        return this.cells;
    }

    public getCell(col: number, row: number): GridCell {

        for (const cell of this.cells) {

            if (cell.relativePosition.x === row && cell.relativePosition.y === col) {
                return cell;
            }
        }

        return undefined;

    }

    public getCellByAbsolutePosition(x: number, y: number): GridCell {

        if (undefined === this.absoluteMap[x]) {
            return undefined;
        }

        if (undefined === this.absoluteMap[x][y]) {
            return undefined;
        }

        const index = this.absoluteMap[x][y];

        return this.cells[index];
    }

}