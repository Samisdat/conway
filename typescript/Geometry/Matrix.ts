import {Position} from '@Conway/Geometry/Position';
import {Boundposition} from '@Conway/Geometry/Boundposition';

interface MatrixInterface {
    [index: string]: Position;
}

interface Dimension {
    width: number;
    height: number;
}

type PositionAttributes = 'x' | 'y';

export class Matrix {

    private matrix: MatrixInterface = {};

    private boundposition: Boundposition = new Boundposition(
        new Position(0, 0),
        new Position(0, 0),
    );

    public has(position: Position): boolean {

        return (undefined !== this.matrix[position.toString()]);

    }

    public add(position: Position): void {

        if (true === this.has(position)) {
            return;
        }

        this.matrix[position.toString()] = position;

        this.boundposition.expand(position);

    }

    public remove(position: Position): void {

        if (false === this.has(position)) {
            return;
        }

        delete this.matrix[position.toString()];

    }

    public all(): Position[] {

        const positions: Position[] = [];

        for (const index in this.matrix) {

            positions.push(this.matrix[index]);

        }

        return positions;

    }

    public getBound(): Boundposition {

        return this.boundposition;

    }

    private getDimension(): Dimension {

        const boundPosition = this.getBound();

        const dimension: Dimension = {
            width: boundPosition.topRight().x - boundPosition.bottomLeft().x,
            height: boundPosition.topRight().y - boundPosition.bottomLeft().y,
        }

        return dimension;

    }

    public width(): number {

        return this.getDimension().width;

    }

    public height(): number {

        return this.getDimension().height;

    }

}
