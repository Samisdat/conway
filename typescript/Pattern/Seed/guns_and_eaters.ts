import {SeedFunction} from '@Conway/Frontend/Web';
import {GridCreator} from '@Conway/Grid/GridCreator';
import {CellMatrix} from '@Conway/Conway/CellMatrix';
import {Position} from '@Conway/Conway/Position';
import {readPatternFromPlainFile} from '@Conway/Pattern/readPatternFromPlainFile';
import {gunsAndEatersPattern} from '@Conway/Pattern/Store/gunsAndEaters.cells';


export const gunsAndEaters: SeedFunction = (gridCreator: GridCreator, matrix: CellMatrix) => {

    const gunsAndEaters = readPatternFromPlainFile(gunsAndEatersPattern);

    const patternGutter = 2;

    let repeat = Math.floor(gridCreator.getRows() / (gunsAndEaters.getWidth() + patternGutter ));

    if (0 === repeat % 2) {
        repeat += 1;
    }

    const patternsPerSide = Math.floor(repeat / 2);

    matrix.seedPattern(
        gunsAndEaters,
        new Position(0, -1)
    );

    let move = new Position(0, -1);

    for (let i = 0; i < patternsPerSide; i += 1) {

        move = move.move(
            new Position(
                -1 * (gunsAndEaters.getWidth() + patternGutter), 0
            )
        );

        matrix.seedPattern(
            gunsAndEaters,
            move
        );

    }

    move = new Position(0, -1);

    for (let i = 0; i < patternsPerSide; i += 1) {

        move = move.move(
            new Position(
                gunsAndEaters.getWidth() + patternGutter, 0
            )
        );

        matrix.seedPattern(
            gunsAndEaters,
            move
        );

    }

};
