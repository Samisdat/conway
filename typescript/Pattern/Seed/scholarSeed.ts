import {SeedFunction} from '@Conway/Frontend/Web';
import {GridCreator} from '@Conway/Grid/GridCreator';
import {CellMatrix} from '@Conway/Conway/CellMatrix';
import {Position} from '@Conway/Conway/Position';
import {readPatternFromPlainFile} from '@Conway/Pattern/readPatternFromPlainFile';
import {scholar} from '@Conway/Pattern/scholar.cells';

export const scholarSeed: SeedFunction = (gridCreator: GridCreator, matrix: CellMatrix) => {

    const pattern = readPatternFromPlainFile(scholar);

    pattern.rotate();

    matrix.seedPattern(
        pattern,
        new Position(-75, 0)
    );

    matrix.seedPattern(
        pattern,
        new Position(75, 0)
    );

};

