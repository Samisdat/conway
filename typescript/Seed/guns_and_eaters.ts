import {GridCreator} from '../Grid/GridCreator';
import {CellMatrix} from '../Conway/CellMatrix';
import {Patterns} from '../Conway/Patterns';
import {Position} from '../Conway/Position';
import {SeedFunction} from '../Frontend/Web';

export const guns_and_eaters: SeedFunction = (gridCreator: GridCreator, matrix: CellMatrix) => {

    const patterns = new Patterns();
    const gunsAndEaters = patterns.get('guns_and_eaters');

    const patternGutter = 2;

    let repeat = Math.floor(gridCreator.getRows() / (gunsAndEaters.getWidth() + patternGutter ));

    if (0 === repeat % 2) {
        repeat += 1;
    }

    let patternsPerSide = Math.floor(repeat / 2);

    matrix.seedPattern(
        patterns.get('guns_and_eaters'),
        new Position(0, -1)
    );


    let move = new Position(0, -1);

    if (0 === repeat % 2) {
        move = move.move(
            new Position(-1 * Math.floor((gunsAndEaters.getWidth() + patternGutter )  / 2 ), 0)
        );
    }

    for (let i = 0; i < patternsPerSide; i += 1) {

        move = move.move(
            new Position(
                -1 * (gunsAndEaters.getWidth() + patternGutter), 0
            )
        );

        matrix.seedPattern(
            patterns.get('guns_and_eaters'),
            move
        );

    }

    move = new Position(0, -1);

    if (0 === repeat % 2) {
        move = move.move(
            new Position(-1 * Math.floor((gunsAndEaters.getWidth() + patternGutter )  / 2 ), 0)
        );
    }

    for (let i = 0; i < patternsPerSide; i += 1) {

        move = move.move(
            new Position(
                gunsAndEaters.getWidth() + patternGutter, 0
            )
        );

        matrix.seedPattern(
            patterns.get('guns_and_eaters'),
            move
        );

    }

};
