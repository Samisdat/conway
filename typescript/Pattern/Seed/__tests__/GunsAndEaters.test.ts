import serializer from '../../../../jest-serialize-conway';
import {Position} from '@Conway/Geometry/Position';
import {CellMatrix} from '@Conway/Geometry/CellMatrix';
import {gunsAndEaters} from '@Conway/Pattern/Seed/guns_and_eaters';
import {GridCreator} from '@Conway/Frontend/Grid/GridCreator';

describe('guns_and_eaters', () => {

    beforeAll(() => {

        expect.addSnapshotSerializer(serializer);

    });

    test('seed with even grid', () => {

        const gridCreator = new GridCreator(
            200,
            100,
            1,
            new Position(0, 0),
            1
        );

        const matrix = new CellMatrix();

        gunsAndEaters(gridCreator, matrix);

        expect(matrix).toMatchSnapshot();

    });

    test('seed with odd grid', () => {

        const gridCreator = new GridCreator(
            50,
            100,
            1,
            new Position(0, 0),
            1
        );

        const matrix = new CellMatrix();

        gunsAndEaters(gridCreator, matrix);

        expect(matrix).toMatchSnapshot();

    });

});
