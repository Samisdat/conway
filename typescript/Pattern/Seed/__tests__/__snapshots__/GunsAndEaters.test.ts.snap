// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`guns_and_eaters seed with even grid 1`] = `
OOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOO
OOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOO
OOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXXOOOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXXOOOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXXOOOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXXOOOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXX
OOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXXOOOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXXOOOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXXOOOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXXOOOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXX
OOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOOOOOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOOOOOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOOOOOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOOOOOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOO
OOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOO
OOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOOOOOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOOOOOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOOOOOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOOOOOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOO
XXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOOOXXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOOOXXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOOOXXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOOOXXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOO
XXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOOOXXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOOOXXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOOOXXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOOOXXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOO
OOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOO
OOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOO
`;

exports[`guns_and_eaters seed with odd grid 1`] = `
OOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOO
OOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOO
OOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXX
OOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXX
OOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOO
OOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOO
OOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOO
XXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOO
XXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOO
OOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOO
OOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOO
`;
